### PR TITLE
Update for Capstone 3.0.5

### PR DESCRIPTION
**Pull request by gwasser**
Capstone 3.0.5-rc2 seems to remove the M68K architecture, and makes a couple of other breaking changes to the X86 and ARM architectures.

I updated the code to match latest Capstone, updated tests, and added a few examples since documentation was hard to find on how to actually use this.

### CODE REVIEW

File: src/Hapstone/Internal/X86.chs [hunk 1]
```diff
@@ -165,20 +162,18 @@ data CsX86 = CsX86
...
-    , eflags :: Word64 -- ^ the EFLAGS set by this instruction
     , operands :: [CsX86Op] -- ^ operand list for this instruction, *MUST*
                             -- have <= 8 elements, else you'll get a runtime
                             -- error when you (implicitly) try to write it to
                             -- memory via it's Storable instance
     } deriving (Show, Eq)
 
 instance Storable CsX86 where
-    sizeOf _ = 448
+    sizeOf _ = 423
```
Comment (ibabushkin) [inline src/Hapstone/Internal/X86.chs]: This is probably incorrect, as the alignment is still specified to be 8. thus, the size must be a multiple of 8 as well. How did you calculate this number?

File: test/Internal/X86/default_sizes [hunk 2]
```diff
@@ -13,7 +13,7 @@ size:		32	1
 avx_bcast:	36	4
 avx_zero_opmask:40	1
 
-cs_x86:	432
+cs_x86:	423
```
Comment (ibabushkin) [inline test/Internal/X86/default_sizes]: Same thing here - this will need some more investigation :)

File: hapstone.cabal [hunk 3]
```diff
@@ -19,7 +19,6 @@ library
                        Hapstone.Internal.Util,
                        Hapstone.Internal.Arm64,
                        Hapstone.Internal.Arm,
-                       Hapstone.Internal.M68K,
```
Comment (ibabushkin) [inline hapstone.cabal:22]: This sounds weird. The `next` branch of the main capstone repo still includes `m68k` and I didn't notice any plans on removing a useful and working chunk of code on the issue tracker.

**Review by ibabushkin — COMMENTED** [review]
Thanks for the work you've done on this, as I lately didn't have much time to spend on this project, it's greatly appreciated. I'd be happy to merge when we sort out the few details I've commented in the diffs.

File: hapstone.cabal (hunk 3 above)
Comment (gwasser) [reply hapstone.cabal:22]: It looks like M68K was removed from at least the tools in November: https://github.com/aquynh/capstone/commit/8853489ea99b5f47be6b9c8fe7aea4740ab6a987

This is why I had removed it, it appears removed from the 3.0.5 release on purpose. The 3.0.5-rc2 tarball doesn't include M68K in the `arch` folder. Do you think that was an error? I don't see it on the `master` branch which is where I think 3.0.5 is sourced.

Maybe the plan is to re-introduce it in the 4.0 release so it's still on the `next` branch?

File: src/Hapstone/Internal/X86.chs (hunk 1 above)
Comment (gwasser) [reply src/Hapstone/Internal/X86.chs]: I calculated this number by adding the sizes of all properties listed in the default_sizes file. I had thought maybe 432 was a typo in the default_sizes file. It was an oversight that 423 isn't a multiple of 8, so I suppose the nearest is now 424? I pushed a commit to change to 424.

File: hapstone.cabal (hunk 3 above)
Comment (gwasser) [reply hapstone.cabal:22]: I asked in a ticket. The official response is that the 3.0 branch has never had M68K, while `next` still does. So I think I was perhaps targeting a different version than you were. Oops. Perhaps you can merge this work into `master` and then add the M68K code back into your `next` branch for 4.0.

Comment (ibabushkin) [reply hapstone.cabal:22]: I see, sorry for the confusion. I'll look how we'll handle this.

**Review by ibabushkin — COMMENTED** [review]
This looks much better :) I'll have a more through look this evening and we'll be good to go! When this code was initially written, I've documented the procedure and subsequent findings [here](https://ibabushkin.github.io/0x03%20-%20A%20Tale%20of%20Struct%20Alignment), it explains the details a bit.

**Comment by ibabushkin** [comment]
Quick update: I'm currently looking through/fixing the tests to avoid further surprises when the code in question gets new changes. At the end of the week, I'll merge this into master (even though some further tests might be useful)

**Comment by ibabushkin** [comment]
Alright, your commits are on master now, so I will close this PR. I think it should work now, but I'd delay the release to hackage, for a few reasons. Apparently the actual release of 3.0.5 isn't yet done, so distributions still ship 3.0.4. Plus, I'd like to do a fuzzing run on all the marshalling code as it is quite fragile and I'm not entirely sure that we didn't miss anything. Thanks for your contribution!

**Comment by gwasser** [comment]
Thanks for checking into this and merging. I plan on using hapstone/capstone for a personal project so I'll file a new PR if I spot anything fishy.

**Comment by ibabushkin** [comment]
On 2017-05-13, Garret Wassermann wrote:
> Thanks for checking into this and merging. I plan on using
> hapstone/capstone for a personal project so I'll file a new PR if I spot anything fishy.
That's very nice! Please file issues for other issues you encounter as
well :)

**Pull request closed without merging**